### PR TITLE
Uncovered expression count as a coverage metric

### PR DESCRIPTION
**Pull request by SebastianKG**
An alternate way of evaluating expression coverage: the maximum allowed count of expressions not covered.

### CODE REVIEW

File: haskell/private/coverage_wrapper.sh.tpl [hunk 1]
```diff
@@ -42,9 +43,15 @@ $binary_path "$@"
...
-if [ $expression_coverage -lt $expected_expression_coverage ]
+covered_expression_percentage=$(grep "expressions used" __hpc_coverage_report | cut -c 1-3)
+if [ $covered_expression_percentage -lt $expected_covered_expressions_percentage ]
 then
-  echo -e "\n==>$ERRORCOLOR Inadequate expression coverage.$CLEARCOLOR Expected $expected_expression_coverage%, but actual coverage was $ERRORCOLOR$(($expression_coverage))%$CLEARCOLOR.\n"
+  echo -e "\n==>$ERRORCOLOR Inadequate expression coverage percentage.$CLEARCOLOR Expected $expected_covered_expressions_percentage%, but actual coverage was $ERRORCOLOR$(($covered_expression_percentage))%$CLEARCOLOR.\n"
+  exit 1
+fi
+uncovered_expression_count=$(grep "expressions used" __hpc_coverage_report | sed s/.*\(//g | cut -f1 -d "/")
+if [ $uncovered_expression_count -gt $expected_uncovered_expression_count ]
```
Comment (ghorn) [inline haskell/private/coverage_wrapper.sh.tpl]: Could you implement a ratcheting mechanism? If you only check greater than then noone will tighten the bounds. What about having both `expected_uncovered_expression_count` and `max_expected_uncovered_expression_count` and then doing something like:
```
# check max allowed
if [ $uncovered_expression_count -gt $max_expected_uncovered_expression_count ]
then
  echo "Too many uncovered expressions..."
  exit 1
fi

# check exact number
if [ $uncovered_expression_count -ne $expected_uncovered_expression_count ]
then
  echo "You have xxx uncovered expressions but expected YYY."

  if [ $uncovered_expression_count -lt $expected_uncovered_expression_count ]
  then
    echo "Nice job! You got rid of uncovered expressions, please reduce the expected number..."
  elif [ $uncovered_expression_count -gt $expected_uncovered_expression_count ]
    echo "Uh oh please don't make test coverage worse...."
  fi

  exit 1
fi
```

File: haskell/haskell.bzl [hunk 2]
```diff
@@ -130,13 +130,21 @@ def _mk_binary_rule(**kwargs):
...
     )
 
-    # Tests have an extra fields regarding expected code coverage percentages.
+    # Tests have an extra fields regarding code coverage.
     if is_test:
         attrs.update({
-            "expected_expression_coverage": attr.int(
-                default = 0,
+            "expected_covered_expressions_percentage": attr.int(
+                default = -1,
```
Comment (guibou) [inline haskell/haskell.bzl]: Is this `-1` documented?

Comment (guibou) [reply haskell/haskell.bzl]: Actually I don't like sentinel values, but the bazel API does not allows us to use something similar to a `Maybe`. However any sentinel value must be documented as much as possible, OR using something more common such as `None` if possible.

File: haskell/private/haskell_impl.bzl [hunk 3]
```diff
@@ -62,7 +62,8 @@ def haskell_binary_impl(ctx):
     return _haskell_binary_common_impl(ctx, is_test = False)
 
 def _should_inspect_coverage(ctx, hs, is_test):
-    return hs.coverage_enabled and is_test and ctx.attr.expected_expression_coverage > 0
+    return hs.coverage_enabled and is_test and (ctx.attr.expected_covered_expressions_percentage != None or
```
Comment (guibou) [inline haskell/private/haskell_impl.bzl]: `None` or `-1`?

File: haskell/haskell.bzl [hunk 4]
```diff
@@ -130,13 +130,21 @@ def _mk_binary_rule(**kwargs):
...
                 doc = "The expected percentage of expressions covered by testing.",
             ),
+            "expected_uncovered_expression_count": attr.int(
+                default = -1,
+                doc = "The expected number of expressions which are not covered by testing.",
+            ),
+            "strict_coverage_analysis": attr.bool(
```
Comment (guibou) [inline haskell/haskell.bzl]: Shouldn't this be `True` by default?

Comment (SebastianKG) [reply haskell/haskell.bzl]: No, I don't think the behaviour described by strict code coverage is what people expect by default. People expect to be setting minimums/maximums, not exact values. There is value, however, in setting exact values if your team is willing to constantly re-set them for every change (namely, it forces awareness and constant improvement of test coverage).

File: haskell/haskell.bzl (hunk 2 above)
Comment (SebastianKG) [reply haskell/haskell.bzl]: Added a comment internally, but I don't think this should be documented to the user. This is the Nothing value, and if the user provides anything it should be a Just.

**Pull request merged**